src/v/rpc/rpc_server.cc

```diff
@@ -104,7 +104,13 @@ rpc_server::send_reply(ss::lw_shared_ptr<server_context_impl> ctx, netbuf buf) {
 
 ss::future<> rpc_server::send_reply_skip_payload(
   ss::lw_shared_ptr<server_context_impl> ctx, netbuf buf) {
-    co_await ctx->conn->input().skip(ctx->get_header().payload_size);
+    try {
+        co_await ctx->conn->input().skip(ctx->get_header().payload_size);
+    } catch (...) {
+        ctx->body_parse_exception(std::current_exception());
+        co_return;
+    }
+    ctx->signal_body_parse();
```

Comment:
Previously this was signaled after the `send_reply()` finished. I imagine this reordering means that we allow waiters to proceed before responding to clients. Is that important to this fix?

Reply:
Not particularly important to the fix. And yeah, it would allow for waiters to continue before the reply is sent. In this case it would allow for more requests to be read from the connection before a reply is sent.

We currently have three code paths in `dispatch_method_once`
1. For unsupported version
2. For unknown methods
3. For known methods

For 1 & 2 we currently only signal waiters after a reply is sent.
For 3 though we signal the waiters right after consuming & parsing all input for the request. Long before calling the service method and sending the reply. 

The goal of this change was to have 1 & 2 match the behavior of 3 as I couldn't see a point of waiting till after a reply to signal the waiters. If you think it's out of scope for this PR though I can start a separate PR for it.

Reply:
I see, thanks for the explanation! It makes sense that we'd be able to / want to signal early. I'm also not sure if this was originally implemented intentionally, so it might be worth poking others familiar in the area, but the rationale for standardizing on #3 makes sense to me

Reply:
CC: @dotnwat

Reply:
> I see, thanks for the explanation! It makes sense that we'd be able to / want to signal early.

i agree with @andrwng that it makes sense, and i think you are doing the right thing here @ballard26.

> would allow for waiters to continue before the reply is sent.

is this optimization the only reason for changing the order of operations?

>  I'm also not sure if this was originally implemented intentionally,

however, i think we need to be very conservative in the rpc server because we've had countless issues with race conditions and haven't had time to revamp/rearch rpc. so anything that potentially alters the scheduling is a red flag. in this case it seems completely benign, but its also a rare error case so the optmiization doesn't seem worth working through the concerns?

Reply:
Fair point about the issues in the past with race conditions. I'll go ahead and revert things to handle the exception with a `finally` in the `rpc_server::dispatch_method_once` in order to preserve the scheduling order. The optimization I made does seem outside the scope of a PR to fix an issue.